include/swift/AST/Module.h

```diff
@@ -1224,12 +1242,29 @@ class LoadedFile : public FileUnit {
     assert(classof(this) && "invalid kind");
   }
 
+  /// A map from private/fileprivate decls to the file they were defined in.
+  llvm::DenseMap<const ValueDecl *, Identifier> FilenameForPrivateDecls;
```

Comment:
Can you explain why this is necessary? It seems like it should be an on-demand on-disk hash table like the private discriminator map.

Reply:
This works in the same way as ``PrivateDiscriminatorsByValue`` when we deserialize a value decl we put it in the map.

When we need to determine ``hasPrivateImport(const swift::ValueDecl *ofDecl)`` we check it.

Reply:
Oops. I didn't realize we did that eagerly for private discriminators too.

…and now I'm working through the reasoning that leads there again: you can't easily look up a decl pointer in a module. My mistake!